calicoctl/resourcemgr/resourcemgr.go

```diff
@@ -56,78 +85,201 @@ type ResourceActionCommand func(*client.Client, unversioned.Resource) (unversion
+	create, update, delete, get, list ResourceActionCommand) {
 
 	if helpers == nil {
-		helpers = make(map[unversioned.TypeMetadata]resourceHelper)
+		helpers = make(map[schema.GroupVersionKind]resourceHelper)
 	}
 
-	tmd := res.GetTypeMetadata()
 	rh := resourceHelper{
-		typeMetadata:      tmd,
+		resource:          res,
 		resourceType:      reflect.ValueOf(res).Elem().Type(),
 		tableHeadings:     tableHeadings,
 		tableHeadingsWide: tableHeadingsWide,
 		headingsMap:       headingsMap,
 		isList:            false,
-		apply:             apply,
+		isNamespaced:      isNamespaced,
 		create:            create,
 		update:            update,
 		delete:            delete,
+		get:               get,
 		list:              list,
 	}
-	helpers[tmd] = rh
+	helpers[res.GetObjectKind().GroupVersionKind()] = rh
 
-	tmd = resList.GetTypeMetadata()
 	rh = resourceHelper{
-		typeMetadata:      tmd,
+		listResource:      resList.(ResourceListObject),
 		resourceType:      reflect.ValueOf(resList).Elem().Type(),
 		tableHeadings:     tableHeadings,
 		tableHeadingsWide: tableHeadingsWide,
 		headingsMap:       headingsMap,
 		isList:            true,
 	}
-	helpers[tmd] = rh
+	helpers[resList.GetObjectKind().GroupVersionKind()] = rh
+
+	for _, v := range names {
+		kindToRes[v] = res
+	}
+}
+
+func (rh resourceHelper) GetObjectType() reflect.Type {
+	return rh.resourceType
+}
+
+// Apply is an un-typed method to apply (create or update) a resource.  This calls directly
+// through to the resource helper specific Apply method which will map the untyped call to
+// the typed interface on the client.
+func (rh resourceHelper) Apply(client client.Interface, resource runtime.Object) (runtime.Object, error) {
+	// Try to create the resource first.
+	ro, err := rh.Create(client, resource)
+
+	switch err.(type) {
+	// Try updating if the resource already exists.
+	case cerrors.ErrorResourceAlreadyExists:
+		return rh.Update(client, resource)
+
+	// For any other errors, return the error
+	default:
+		return ro, err
+	}
+}
+
+// Create is an un-typed method to create a new resource.  This calls directly
+// through to the resource helper specific Create method which will map the untyped call to
+// the typed interface on the client.
+func (rh resourceHelper) Create(client client.Interface, resource runtime.Object) (runtime.Object, error) {
+	return rh.create(client, resource.(ResourceObject))
+}
+
+// Update is an un-typed method to update an existing resource.  This calls directly
+// through to the resource helper specific Update method which will map the untyped call to
+// the typed interface on the client.
+func (rh resourceHelper) Update(client client.Interface, resource runtime.Object) (runtime.Object, error) {
+	var err error
+	for i := 0; i < 5; i++ {
+		ro, err := rh.GetOrList(client, resource.(ResourceObject))
+		if err != nil {
+			return ro, err
+		}
+		if resource.(ResourceObject).GetObjectMeta().GetName() != "" {
```

Comment:
Does this work?  Shouldn't you be checking if this is a ResourceObject or a ResourceList *before* casting it?  Won't this panic when it does a resource list?

Reply:
Actually, shouldn't the above call directly into the private Get() function rather than using the GetOrList?